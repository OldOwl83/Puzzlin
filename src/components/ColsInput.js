import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
