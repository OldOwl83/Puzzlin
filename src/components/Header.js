import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
