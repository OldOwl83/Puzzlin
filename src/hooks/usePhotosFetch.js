/*
* It watches the URL for query parameters and handles the fetching and storing of data for each search. If a search is repeated, it falls back to localStorage in order to save API requests. It returns a state object containing the array of fetched "photos" and the "loading" status (true or false). If there was an error in the fetch, the "photos" property becomes false.
*/

import { useState, useEffect } from 'react';
import { useLocation } from 'react-router-dom';

import { photosFetch } from '../functions/photosFetch';

export const usePhotosFetch = () => {
    // console.log("AquÃ­ usephotosFetch!!");

    const query = useLocation().search.slice(3);
    
    const [ searchState, setSearchState ] = useState( 
        {
            photos: [],
            loading: true,
        } );
    
    useEffect(() => {

        if( !query )
    
            setSearchState( { photos: [ 
                { 
                    id: 'portada1', 
                    description: 'Imagen de portada 1.', 
                    urlSmall: 'images/portada1.png', 
                    urlRegular: 'images/portada1.png', 
                    author: 'Mauro Donnantuoni' 
                },
                { 
                    id: 'portada2', 
                    description: 'Imagen de portada 2.', 
                    urlSmall: 'images/portada2.png', 
                    urlRegular: 'images/portada2.png', 
                    author: 'Mauro Donnantuoni' 
                },
                { 
                    id: 'portada3', 
                    description: 'Imagen de portada 3.', 
                    urlSmall: 'images/portada3.png', 
                    urlRegular: 'images/portada3.png', 
                    author: 'Mauro Donnantuoni' 
                },
                { 
                    id: 'portada4', 
                    description: 'Imagen de portada 4.', 
                    urlSmall: 'images/portada4.png', 
                    urlRegular: 'images/portada4.png', 
                    author: 'Mauro Donnantuoni' 
                }, 
            
            ], loading: false} );
       
        else if( localStorage.getItem( `puzzlin-${query}` ))

            setSearchState( { 
                photos: JSON.parse( localStorage.getItem( `puzzlin-${query}` ) ), 
                loading: false,
             } );

        else
        {
            setSearchState( { photos: [], loading: true } );

            photosFetch( query )
                .then( phs => 
                    { 
                        setSearchState( { photos: phs, loading: false} )
                       
                        if( query )
                            localStorage.setItem( `puzzlin-${query}`, JSON.stringify( phs ) );
                    })
                .catch( ( err ) => 
                    {
                        console.log("No photos could be retrieved from the server. Error: ", err);
                        setSearchState( { photos: false, loading: false})
                    });                   
        }
        
    }, [ setSearchState, query ]);
    
    return searchState;
};
