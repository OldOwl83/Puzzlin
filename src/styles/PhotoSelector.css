h2{
    color: rgb(18, 112, 116);
    font-size: 1.3rem;
}

#selectorContainer{
    display: flex;
    justify-content:space-between;

    align-items: center;

    width: 96%;
    margin: 6px auto;
    padding: 14px;
}

.message{
    display: inline-block;

    width: 66%;

    font-size: 2.2rem;
    text-align: center;

    color: rgb(45, 22, 88);
}

#loading{
    font-size: 3rem;
    
    transform-origin: center;
    animation: turn 1.4s infinite;
}

@keyframes turn{
    from { transform: rotate( 0 ); }
    to { transform: rotate( 360deg ); }
}

