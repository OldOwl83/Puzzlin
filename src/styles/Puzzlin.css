h1{
    color: rgb(1, 80, 83);
    font-size: 2.4rem;
    padding: 0 40px;
}

hr{
    border-color: beige;
}