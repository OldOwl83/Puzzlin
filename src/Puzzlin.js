import { 
    BrowserRouter as Router,
    Routes, Route,
        } from "react-router-dom"; 

import { createContext, useState } from "react";
import { PhotoSelector } from "./components/PhotoSelector";
import { Puzzle } from "./components/Puzzle";
import { NoRoute } from "./components/NoRoute";

import './styles/Puzzlin.css';

export const MainPhotoContext = createContext( null );
export const PuzzleGrid = createContext( [] );

export const Puzzlin = () => {

    const [ puzzlePhoto, setPuzzlePhoto ] = useState( '' );
    const [ puzzleGrid, setPuzzleGrid ] = useState( [4, 3] );

    return (
        <>
            <h1>Puzzlin</h1>
            <hr />

            <Router>
                <MainPhotoContext.Provider value={ { puzzlePhoto, setPuzzlePhoto } } >
                <PuzzleGrid.Provider value={ { puzzleGrid, setPuzzleGrid } }>

                    <Routes>

                        <Route path="/" element={ <PhotoSelector /> } />
                        <Route path="/?q=:query" element={ <PhotoSelector /> } />

                        <Route path="/puzzle" element={ <Puzzle /> } />

                        <Route path="/*" element={ <NoRoute /> } />

                    </Routes>

                </PuzzleGrid.Provider>
                </MainPhotoContext.Provider>
            </Router>

        </>
    )
};