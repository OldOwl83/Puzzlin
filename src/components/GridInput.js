/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
