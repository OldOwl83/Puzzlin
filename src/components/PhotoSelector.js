import { useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";

import { usePhotosFetch } from "../hooks/usePhotosFetch";
import { InputSubject } from "./InputSubject";
import { PhotoShowcase } from "./PhotoShowcase";
import { GridInput } from "./GridInput";
import { Preview } from './Preview';

import '../styles/PhotoSelector.css';
import 'animate.css';

export const PhotoSelector = () => {

    // console.log('PhotoSelector');
    const location = useLocation();
    const navigate = useNavigate();

    const { photos, loading } = usePhotosFetch();

    useEffect(() => {
        
        if(photos.length > 0 && location.state === null )
        {
                navigate( location, {
                    replace: true,
                    state: {
                        selectedPhoto: photos[0].urlRegular,
                        contScroll: 0,
                    }
                });
        }
            
    }, [ photos ]);

    return (
        <>
            <InputSubject />

            <div id="selectorContainer">
                
                <PhotoShowcase photos={ photos } />

                { loading && <p className="message"><i id="loading" className="fas fa-spinner"></i></p> }
                    
                { !photos && <p className="message">Connection to images provider failed.</p> }

                { !loading && photos.length < 1 && <p className="message">There were no matches with the search keywords.</p> }

                { !loading && photos.length > 0 && 
                
                <div id="previewContainer">
                    
                    <GridInput />
                    <Preview /> 
                    
                </div>}
                    
            </div>
        </>
    )
}