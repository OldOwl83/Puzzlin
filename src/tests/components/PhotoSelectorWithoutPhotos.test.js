import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import '../../functions/photosFetch';
import '../../hooks/usePhotosFetch';
import { PhotoSelector } from '../../components/PhotoSelector';
import { MainPhotoContext, PuzzleGrid } from '../../Puzzlin';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
    useLocation: () => { return { search: '?q=car', state: null }; },
}));

const photos = [];

jest.mock( '../../hooks/usePhotosFetch', () => {
    return {
        usePhotosFetch: () => { return {photos: photos, loading: false}; },
    };
});


describe('Tests on <PhotoSelector /> component.', () => {
  
    const wrapper = mount( 

        <MemoryRouter>
        <MainPhotoContext.Provider value={ { setPuzzlePhoto: jest.fn(), } }>
        <PuzzleGrid.Provider value={ { puzzleGrid: [ 4, 3 ], setPuzzleGrid: jest.fn(), } }>

            <PhotoSelector /> 

        </PuzzleGrid.Provider>
        </MainPhotoContext.Provider>
        </MemoryRouter>
    );

    test('It should match the snapshot and display the "previewContainer" element.', () => {
      
        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find( '#previewContainer' ).exists()).toBe( false );
        expect( wrapper.find( '.message' ).exists()).toBe( true );
        expect( wrapper.find( '.message' ).text()).toBe( 'There were no matches with the search keywords.' );
    });
});
