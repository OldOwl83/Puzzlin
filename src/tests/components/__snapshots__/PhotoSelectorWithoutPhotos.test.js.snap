// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <PhotoSelector /> component. It should match the snapshot and display the "previewContainer" element. 1`] = `
Array [
  <h2>
    Input a subject and select a photo. Then, press on the preview to start.
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="on"
        id="inputSubject"
        onChange={[Function]}
        type="text"
        value=""
      />
    </form>
  </h2>,
  <div
    id="selectorContainer"
  >
    <div
      id="sampler"
    >
      <div
        className="material-icons arrows"
        onClick={[Function]}
      >
        keyboard_double_arrow_up
      </div>
      <div
        id="photosHidder"
      >
        <div
          id="photosContainer"
          style={
            Object {
              "height": 0,
              "top": 0,
            }
          }
        />
      </div>
      <div
        className="material-icons arrows"
        onClick={[Function]}
      >
        keyboard_double_arrow_down
      </div>
    </div>
    <p
      className="message"
    >
      There were no matches with the search keywords.
    </p>
  </div>,
]
`;
